api/src/main/java/jakarta/data/Restrict.java

```diff
@@ -0,0 +1,232 @@
+    }
+
+    public static <T> TextRestriction<T> notContains(String substring, String field) {
+        String pattern = toLikeEscaped(CHAR_WILDCARD, STRING_WILDCARD, true, substring, true);
+        return new TextRestrictionRecord<>(field, NOT, Operator.LIKE, ESCAPED, pattern);
+    }
+
+    public static <T> TextRestriction<T> notEndsWith(String suffix, String field) {
+        String pattern = toLikeEscaped(CHAR_WILDCARD, STRING_WILDCARD, true, suffix, false);
+        return new TextRestrictionRecord<>(field, NOT, Operator.LIKE, ESCAPED, pattern);
+    }
+
+    public static <T> Restriction<T> notIn(Set<Object> values, String field) {
+        return new BasicRestrictionRecord<>(field, NOT, Operator.IN, values);
+    }
+
+    public static <T> TextRestriction<T> notLike(String pattern, String field) {
+        return new TextRestrictionRecord<>(field, NOT, Operator.LIKE, pattern);
+    }
+
+    public static <T> TextRestriction<T> notLike(String pattern,
+                                                  char charWildcard,
+                                                  char stringWildcard,
+                                                  String field) {
+        String p = toLikeEscaped(charWildcard, stringWildcard, false, pattern, false);
+        return new TextRestrictionRecord<>(field, NOT, Operator.LIKE, ESCAPED, p);
+    }
+
+    public static <T> TextRestriction<T> notStartsWith(String prefix, String field) {
+        String pattern = toLikeEscaped(CHAR_WILDCARD, STRING_WILDCARD, false, prefix, true);
+        return new TextRestrictionRecord<>(field, NOT, Operator.LIKE, ESCAPED, pattern);
+    }
+
+    public static <T> TextRestriction<T> startsWith(String prefix, String field) {
+        String pattern = toLikeEscaped(CHAR_WILDCARD, STRING_WILDCARD, false, prefix, true);
+        return new TextRestrictionRecord<>(field, Operator.LIKE, ESCAPED, pattern);
+    }
+
+    /**
+     * Converts the literal pattern into an escaped LIKE pattern.
+     * This method prepends a % character if previous characters are allowed,
+     * escapes the charWildcard (typically _), the stringWildcard (typically %),
+     * and the \ character within the literal by inserting \ prior to each,
+     * and then appends a % character if subsequent characters are allowed.
+     *
+     * @param charWildcard    single character wildcard, typically _.
+     * @param stringWildcard  0 or more character wildcard, typically %.
+     * @param allowPrevious   whether to allow characters prior to the text.
+     * @param literal text    that is not escaped that must be matched.
+     * @param allowSubsequent whether to allow more characters after the text.
+     * @return escaped pattern.
+     * @throws IllegalArgumentException if the same character is supplied for
+     *                                  both wildcard types.
+     */
+    private static String toLikeEscaped(char charWildcard,
```

Comment:
Once everybody agreed with Gavin's `Pattern` class, why not centralize this behavior in that class?

Reply:
Yes, I think it's much better to have a nice class which models this basic concept. It's a much semantically stronger construct.

Reply:
I agree, if Pattern gets added, this utility method can move to there.  Also note that the method is private, so the documentation on it is only for our own benefit and will not show in the API, so users won't even notice once it is moved.